In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker

In [2]:
from mlxtend.frequent_patterns import association_rules, apriori

In [3]:
import pickle
with open("C:/Users/tldms/jupyter/result.pickle","rb") as fr:
    result=pickle.load(fr)

In [4]:

with open("C:/Users/tldms/jupyter/RFML.pickle","rb") as fr:
    RFML=pickle.load(fr)

In [5]:
for a in range(2):
    for b in range (2):
        for c in range(2):
            for d in range(3):
                globals()['group'+str(a)+str(b)+str(c)+str(d)] = RFML[(RFML["Recency_R"] == a) & (RFML["Frequency_F"] == b) & (RFML["Monetary_M"] == c) & (RFML["LPAY_USE"] == d) ]
                globals()['cust'+str(a)+str(b)+str(c)+str(d)] = globals()['group'+str(a)+str(b)+str(c)+str(d)]['CUST']
                globals()['transaction'+str(a)+str(b)+str(c)+str(d)] = pd.merge(result, globals()['cust'+str(a)+str(b)+str(c)+str(d)])

In [6]:
cust1010.head()

5      M000261625
6      M000350564
89     M003641896
100    M003948947
104    M004292419
Name: CUST, dtype: object

In [7]:
transaction1010.head()

CUST  MA_FEM_DV  AGES  C_ZON_HLV  BUY_AM  COP_C  CHNL_DV  PD_C  \
0  M117855953          0    30         11  759500    110        1   949   
1  M117855953          0    30         11  167000    110        1   816   
2  M117855953          0    30         11   48000    110        1   816   
3  M117855953          0    30         11  126500    110        1   959   
4  M117855953          0    30         11   19700    110        1  1227   

   BUY_CT  S_ZON_HLV       DE_DT  MONTH  DATE  DAY  DE_HR CLAC_HLV_NM  \
0       1         11  2021-01-01      1     1    4     11        여성의류   
1       1         11  2021-01-13      1    13    2     20       스포츠패션   
2       1         11  2021-01-24      1    24    6     13       스포츠패션   
3       1         11  2021-02-14      2    14    6     18        여성의류   
4       1         11  2021-02-14      2    14    6     20       자동차용품   

  CLAC_MCLS_NM       PD_NM  LPAY_USE  
0       여성의류상의     여성티셔츠/탑         0  
1       여성스포츠화  여성런닝/트레이닝화         0  
2       여성스포츠화  여성런닝/트레이닝화         0  
3      여성의류아우터        여성재킷         0  
4     액세서리편의용품     수납/정리용품         0

## 큰손nl의 중분류 연관규칙찾는 알고리즘

큰손nl의 구매물품에 대한 연관규칙을 찾기위해서 '중분류'를 기준으로 데이터를 파악한다.<br>
중분류 카테고리 순위만 파악한 다음 대분류로 알고리즘을 돌려볼 것이다.<br>
큰손nl은 rfml기준으로 1010이라서 1010의 그룹에 속한 cust의 transaction데이터로 분석을 진행한다.<br>

여기서 cop_u도 포함되어있으니 cop_u인것을 버리기위해 조건을 건다

In [8]:
transaction1010 = transaction1010[(transaction1010['COP_C']<200)]
transaction1010

CUST  MA_FEM_DV  AGES  C_ZON_HLV  BUY_AM  COP_C  CHNL_DV  PD_C  \
0       M117855953          0    30         11  759500    110        1   949   
1       M117855953          0    30         11  167000    110        1   816   
2       M117855953          0    30         11   48000    110        1   816   
3       M117855953          0    30         11  126500    110        1   959   
4       M117855953          0    30         11   19700    110        1  1227   
...            ...        ...   ...        ...     ...    ...      ...   ...   
131128  M058847456          0    40         10   59000    160        2   969   
131129  M058847456          0    40         10   39200    160        2   290   
131130  M058847456          0    40         10  158000    160        2  1718   
131131  M058847456          0    40         10   10800    160        2  1717   
131132  M058847456          0    40         10   26000    160        2   954   

        BUY_CT  S_ZON_HLV       DE_DT  MONTH  DATE  DAY  DE_HR CLAC_HLV_NM  \
0            1         11  2021-01-01      1     1    4     11        여성의류   
1            1         11  2021-01-13      1    13    2     20       스포츠패션   
2            1         11  2021-01-24      1    24    6     13       스포츠패션   
3            1         11  2021-02-14      2    14    6     18        여성의류   
4            1         11  2021-02-14      2    14    6     20       자동차용품   
...        ...        ...         ...    ...   ...  ...    ...         ...   
131128       1          0  2021-05-29      5    29    5     11        여성의류   
131129       1          0  2021-07-02      7     2    4      4        남성의류   
131130       1          0  2021-07-29      7    29    3     18        패션잡화   
131131       1          0  2021-10-23     10    23    5     18        패션잡화   
131132       1          0  2021-12-01     12     1    2     16        여성의류   

       CLAC_MCLS_NM       PD_NM  LPAY_USE  
0            여성의류상의     여성티셔츠/탑         0  
1            여성스포츠화  여성런닝/트레이닝화         0  
2            여성스포츠화  여성런닝/트레이닝화         0  
3           여성의류아우터        여성재킷         0  
4          액세서리편의용품     수납/정리용품         0  
...             ...         ...       ...  
131128       여성의류하의       여성청바지         0  
131129       남성의류상의       남성티셔츠         0  
131130          남성화       남성정장화         0  
131131          남성화       남성슬리퍼         0  
131132       여성의류상의   여성스웨터/풀오버         0  

[120701 rows x 19 columns]

association rule mining에 필요한 컬럼인 'CUST','CLAC_MCLS_NM','DE_DT'를 가져와서 분석한다

In [9]:
transaction1010_m = transaction1010[['CUST','CLAC_MCLS_NM','DE_DT']]
transaction1010_m.head()

CUST CLAC_MCLS_NM       DE_DT
0  M117855953       여성의류상의  2021-01-01
1  M117855953       여성스포츠화  2021-01-13
2  M117855953       여성스포츠화  2021-01-24
3  M117855953      여성의류아우터  2021-02-14
4  M117855953     액세서리편의용품  2021-02-14

In [17]:
a = transaction1010.groupby(['CUST','DE_DT']).CLAC_MCLS_NM.unique().reset_index()
a['transaction'] = a.index
a = a.drop(['CUST','DE_DT'], axis=1)
a = a[['transaction','CLAC_MCLS_NM']]
a['count'] = 1
a

transaction                                       CLAC_MCLS_NM  count
0                0                                             [스킨케어]      1
1                1                       [패션액세서리, 바디케어, 식당, 여성속옷, 와인]      1
2                2                                               [식당]      1
3                3                                       [스킨케어, 메이크업]      1
4                4                                             [여성속옷]      1
...            ...                                                ...    ...
35530        35530  [여성지갑, 남성지갑, 남성스포츠화, 여성등산/아웃도어의류, 여성가방, 여성의류상의...      1
35531        35531                [남성등산/아웃도어의류, 요구르트, 봉지라면, 맥주, 유가공품]      1
35532        35532                                      [남성등산/아웃도어의류]      1
35533        35533                             [남성일반스포츠의류, 생활서비스, 식당]      1
35534        35534                                  [식당, 탄산음료, 기능성음료]      1

[35535 rows x 3 columns]

In [18]:
transaction = pd.DataFrame({'Transaction' : [], 'Item' : [], 'Count' : []})
num = 0
for i in range(35535):
    for count in range(len(a.iloc[i,1])):
        transaction.loc[num] = [i,a.iloc[i,1][count],1]
        num+=1
transaction
# 

Transaction    Item  Count
0                0    스킨케어      1
1                1  패션액세서리      1
2                1    바디케어      1
3                1      식당      1
4                1    여성속옷      1
...            ...     ...    ...
87530        35533   생활서비스      1
87531        35533      식당      1
87532        35534      식당      1
87533        35534    탄산음료      1
87534        35534   기능성음료      1

[87535 rows x 3 columns]

위의 transaction의 거래내역을 각 행으로 변환하는 작업의 반복문 코드가 너무 오래걸려서 <br>
pickle 파일로 변환시켰다.

In [20]:
import pickle
with open("transaction_bhnl_m.pickle","wb") as fw:
    pickle.dump(transaction, fw)

In [8]:
with open("C:/Users/tldms/jupyter/transaction_bhnl_m.pickle","rb") as fr:
    transaction=pickle.load(fr)

In [9]:
basket = transaction.pivot_table(index='Transaction', columns='Item', values='Count', aggfunc='sum').fillna(0)
basket.head()

Item         DVD/홈시어터   TV  가공유  가방세트  간식/안주형과자  간편조리  갑각류  거실가구  건강가전  \
Transaction                                                              
0                 0.0  0.0  0.0   0.0       0.0   0.0  0.0   0.0   0.0   
1                 0.0  0.0  0.0   0.0       0.0   0.0  0.0   0.0   0.0   
2                 0.0  0.0  0.0   0.0       0.0   0.0  0.0   0.0   0.0   
3                 0.0  0.0  0.0   0.0       0.0   0.0  0.0   0.0   0.0   
4                 0.0  0.0  0.0   0.0       0.0   0.0  0.0   0.0   0.0   

Item         건강관리용품  ...  핸드/풋케어  햄/소시지   향수  향신료  헤어액세서리  헤어케어  호주산소고기  홈웨어  \
Transaction          ...                                                       
0               0.0  ...     0.0    0.0  0.0  0.0     0.0   0.0     0.0  0.0   
1               0.0  ...     0.0    0.0  0.0  0.0     0.0   0.0     0.0  0.0   
2               0.0  ...     0.0    0.0  0.0  0.0     0.0   0.0     0.0  0.0   
3               0.0  ...     0.0    0.0  0.0  0.0     0.0   0.0     0.0  0.0   
4               0.0  ...     0.0    0.0  0.0  0.0     0.0   0.0     0.0  0.0   

Item         홍삼/인삼가공식품  화장지/티슈  
Transaction                     
0                  0.0     0.0  
1                  0.0     0.0  
2                  0.0     0.0  
3                  0.0     0.0  
4                  0.0     0.0  

[5 rows x 330 columns]

In [10]:
freq_item = apriori(basket, min_support = 0.005,use_colnames = True)
freq_item

C:\Users\tldms\anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support          itemsets
0    0.011060             (가공유)
1    0.009005              (건면)
2    0.009427             (건어물)
3    0.007767             (견과류)
4    0.016238             (계란류)
..        ...               ...
154  0.005459  (여성의류하의, 여성의류상의)
155  0.008499       (열매채소, 잎채소)
156  0.005600        (우유, 요구르트)
157  0.005572        (우유, 청소용품)
158  0.006107       (청소용품, 잎채소)

[159 rows x 2 columns]

In [23]:
freq_item.sort_values('support', ascending = False, inplace = True)
pd.set_option('display.max_rows', 500000)
freq_item

support           itemsets
55   0.176333               (식당)
68   0.071141           (여성의류상의)
70   0.060785          (여성의류아우터)
106  0.059772             (푸드코트)
53   0.056508              (스낵류)
8    0.051921             (국산과일)
18   0.045167           (남성의류상의)
19   0.042437           (남성의류세트)
22   0.042184        (남성일반스포츠의류)
43   0.039679             (비스킷류)
87   0.033094             (임대매출)
98   0.032728             (청소용품)
54   0.032025             (스킨케어)
17   0.029858           (남성스포츠화)
52   0.029520             (수입과일)
69   0.029267           (여성의류세트)
40   0.027635             (베이커리)
34   0.027494             (레토르트)
83   0.027016               (우유)
74   0.025918              (여성화)
88   0.025552              (잎채소)
104  0.024933           (패션액세서리)
81   0.023948             (요구르트)
62   0.023864             (여성가방)
29   0.022626             (냉장조리)
44   0.021866                (빵)
31   0.021556          (두부/콩나물류)
36   0.021303             (메이크업)
77   0.021190             (열매채소)
59   0.020768              (양채류)
100  0.020628             (커피음료)
42   0.019924             (봉지라면)
101  0.019474             (탄산음료)
35   0.018770               (맥주)
65   0.018545             (여성속옷)
26   0.018151            (냉동간편식)
61   0.018039            (어묵/맛살)
103  0.017504          (파티/팬시용품)
20   0.017504          (남성의류아우터)
16   0.017476      (남성등산/아웃도어의류)
48   0.016800               (생수)
72   0.016463           (여성의류하의)
4    0.016238              (계란류)
46   0.016069            (사탕/캔디)
23   0.015675              (남성화)
11   0.015196            (그릇/식기)
27   0.015056            (냉장간편식)
15   0.014690           (남성골프의류)
50   0.014352              (소스류)
63   0.014324           (여성골프의류)
14   0.014155        (기타아웃도어/레저)
39   0.014071              (버섯류)
49   0.013902             (성인침구)
85   0.013592             (유가공품)
86   0.013564             (일반담배)
109  0.013255            (햄/소시지)
105  0.013198             (포장반찬)
6    0.013029             (과채음료)
150  0.012861         (푸드코트, 식당)
5    0.012804               (골프)
96   0.012748       (주방정리용품/소모품)
32   0.012720              (디저트)
80   0.012635               (와인)
92   0.012579             (조리기구)
99   0.012270              (초콜릿)
25   0.011960           (남아의류상의)
10   0.011932            (국산소고기)
73   0.011847        (여성일반스포츠의류)
21   0.011538           (남성의류하의)
137  0.011538        (스낵류, 비스킷류)
60   0.011510               (어류)
71   0.011425           (여성의류전신)
41   0.011088               (보석)
0    0.011060              (가공유)
115  0.010947       (수입과일, 국산과일)
82   0.010778             (용기라면)
12   0.010750            (기능성음료)
57   0.010637            (아이스크림)
51   0.010131               (소주)
28   0.010103             (냉장음료)
148  0.009878       (식당, 여성의류상의)
2    0.009427              (건어물)
66   0.009427           (여성스포츠화)
45   0.009230             (뿌리채소)
79   0.009174             (온장조리)
143  0.009118          (스낵류, 식당)
1    0.009005               (건면)
9    0.008977           (국산돼지고기)
153  0.008808  (여성의류아우터, 여성의류상의)
94   0.008724             (조미료류)
24   0.008724             (남아완구)
110  0.008696               (향수)
149  0.008696      (여성의류아우터, 식당)
7    0.008527             (교육완구)
97   0.008499              (차음료)
155  0.008499        (잎채소, 열매채소)
38   0.008414             (바디케어)
84   0.008358             (원두커피)
58   0.008245              (양념육)
90   0.008161               (장류)
123  0.008161       (청소용품, 국산과일)
78   0.008077              (영양제)
126  0.008020       (식당, 남성의류세트)
30   0.007908             (닭고기류)
76   0.007851           (여아의류상의)
3    0.007767              (견과류)
111  0.007767             (헤어케어)
56   0.007711             (식용유류)
127  0.007683    (남성일반스포츠의류, 식당)
146  0.007654        (청소용품, 스낵류)
116  0.007626        (스낵류, 국산과일)
122  0.007486        (잎채소, 국산과일)
124  0.007457       (남성의류상의, 식당)
117  0.007373         (식당, 국산과일)
95   0.007345             (주방가전)
112  0.007204           (화장지/티슈)
119  0.007204       (국산과일, 열매채소)
108  0.007120              (해조류)
144  0.006669        (스낵류, 요구르트)
118  0.006641        (양채류, 국산과일)
102  0.006501

In [24]:
rules_h = association_rules(freq_item, metric = "lift", min_threshold = 1)
rules_h.sort_values('confidence', ascending = False, inplace = True)
rules_h.head(200)

antecedents  consequents  antecedent support  consequent support  \
37        (버섯류)        (잎채소)            0.014071            0.025552   
9        (열매채소)        (잎채소)            0.021190            0.025552   
4        (수입과일)       (국산과일)            0.029520            0.051921   
74        (버섯류)       (열매채소)            0.014071            0.021190   
23       (열매채소)       (국산과일)            0.021190            0.051921   
8         (잎채소)       (열매채소)            0.025552            0.021190   
68     (여성의류하의)     (여성의류상의)            0.016463            0.071141   
73      (사탕/캔디)        (스낵류)            0.016069            0.056508   
26        (양채류)       (국산과일)            0.020768            0.051921   
31       (봉지라면)        (스낵류)            0.019924            0.056508   
53         (맥주)        (스낵류)            0.018770            0.056508   
28    (두부/콩나물류)       (국산과일)            0.021556            0.051921   
20        (잎채소)       (국산과일)            0.025552            0.051921   
34    (두부/콩나물류)        (잎채소)            0.021556            0.025552   
3        (비스킷류)        (스낵류)            0.039679            0.056508   
39       (열매채소)    (두부/콩나물류)            0.021190            0.021556   
45        (양채류)        (잎채소)            0.020768            0.025552   
38    (두부/콩나물류)       (열매채소)            0.021556            0.021190   
25       (요구르트)        (스낵류)            0.023948            0.056508   
71       (탄산음료)        (스낵류)            0.019474            0.056508   
56        (양채류)       (열매채소)            0.020768            0.021190   
57       (열매채소)        (양채류)            0.021190            0.020768   
59    (두부/콩나물류)       (청소용품)            0.021556            0.032728   
32       (요구르트)       (국산과일)            0.023948            0.051921   
79       (봉지라면)       (청소용품)            0.019924            0.032728   
10       (청소용품)       (국산과일)            0.032728            0.051921   
35        (잎채소)    (두부/콩나물류)            0.025552            0.021556   
75       (열매채소)        (버섯류)            0.021190            0.014071   
36        (잎채소)        (버섯류)            0.025552            0.014071   
41        (잎채소)       (청소용품)            0.025552            0.032728   
44        (잎채소)        (양채류)            0.025552            0.020768   
16       (청소용품)        (스낵류)            0.032728            0.056508   
63       (요구르트)         (우유)            0.023948            0.027016   
46         (우유)        (스낵류)            0.027016            0.056508   
0        (푸드코트)         (식당)            0.059772            0.176333   
49       (베이커리)         (식당)            0.027635            0.176333   
54         (우유)       (국산과일)            0.027016            0.051921   
5        (국산과일)       (수입과일)            0.051921            0.029520   
62         (우유)       (요구르트)            0.027016            0.023948   
64         (우유)       (청소용품)            0.027016            0.032728   
2         (스낵류)       (비스킷류)            0.056508            0.039679   
77        (잎채소)       (수입과일)            0.025552            0.029520   
61       (수입과일)        (스낵류)            0.029520            0.056508   
13     (남성의류세트)         (식당)            0.042437            0.176333   
67       (수입과일)       (청소용품)            0.029520            0.032728   
40       (청소용품)        (잎채소)            0.032728            0.025552   
14  (남성일반스포츠의류)         (식당)            0.042184            0.176333   
58       (청소용품)    (두부/콩나물류)            0.032728            0.021556   
76       (수입과일)        (잎채소)            0.029520            0.025552   
65       (청소용품)         (우유)            0.032728            0.027016   
66       (청소용품)       (수입과일)            0.032728            0.029520   
11       (국산과일)       (청소용품)            0.051921            0.032728   
78       (청소용품)       (봉지라면)            0.032728            0.019924   
51       (비스킷류)       (국산과일)            0.039679            0.051921   
19       (국산과일)        (스낵류)            0.05

## 큰손nl의 대분류 연관규칙 찾는 알고리즘

큰손 NL에 대한 transaction으로 ARM을 진행해본 결과 VIP NL의 고객들이 주로 소비하는 품목들은 <br>
식품(채소, 스낵류 등) 이어서 품목들끼리 비슷해 추천이 의미가 없어 중분류로 마켓팅하는 것은 한계가 있다고 판단해<br>
조금 더 넓은 범위인 대분류를 이용하여 마케팅 전략을 세우려고 한다.

association rule mining에 필요한 컬럼인 'CUST','CLAC_HLV_NM','DE_DT'를 가져와서 분석한다

In [27]:
transaction1010_h = transaction1010[['CUST','CLAC_HLV_NM','DE_DT']]
transaction1010_h

CUST CLAC_HLV_NM       DE_DT
0       M117855953        여성의류  2021-01-01
1       M117855953       스포츠패션  2021-01-13
2       M117855953       스포츠패션  2021-01-24
3       M117855953        여성의류  2021-02-14
4       M117855953       자동차용품  2021-02-14
...            ...         ...         ...
131135  M058847456        NULL  2021-08-14
131136  M058847456        NULL  2021-10-14
131137  M058847456        NULL  2021-11-09
131138  M058847456        NULL  2021-09-24
131139  M058847456        NULL  2021-04-20

[131140 rows x 3 columns]

In [28]:
transaction1010_h = transaction1010_h.groupby(['CUST','DE_DT']).CLAC_HLV_NM.unique().reset_index()
transaction1010_h['transaction'] = transaction1010_h.index
transaction1010_h = transaction1010_h.drop(['CUST','DE_DT'], axis=1)
transaction1010_h = transaction1010_h[['transaction','CLAC_HLV_NM']]
transaction1010_h['count'] = 1
transaction1010_h

transaction                                  CLAC_HLV_NM  count
0                0                                   [화장품/뷰티케어]      1
1                1  [패션잡화, 퍼스널케어, 테넌트/음식점, 속옷/양말/홈웨어, 주류, NULL]      1
2                2                                    [테넌트/음식점]      1
3                3                                   [화장품/뷰티케어]      1
4                4                                  [속옷/양말/홈웨어]      1
...            ...                                          ...    ...
42145        42145                                [스포츠패션, NULL]      1
42146        42146             [스포츠패션, 생활/렌탈서비스, 테넌트/음식점, NULL]      1
42147        42147                                       [NULL]      1
42148        42148                                [테넌트/음식점, 음료]      1
42149        42149                                       [NULL]      1

[42150 rows x 3 columns]

In [29]:
transaction1010_h_df = pd.DataFrame({'Transaction' : [], 'Item' : [], 'Count' : []})
num = 0
for i in range(42150):
    for count in range(len(transaction1010_h.iloc[i,1])):
        transaction1010_h_df.loc[num] = [i,transaction1010_h.iloc[i,1][count],1]
        num+=1
transaction1010_h_df

Transaction       Item  Count
0                0   화장품/뷰티케어      1
1                1       패션잡화      1
2                1      퍼스널케어      1
3                1    테넌트/음식점      1
4                1  속옷/양말/홈웨어      1
...            ...        ...    ...
84254        42146       NULL      1
84255        42147       NULL      1
84256        42148    테넌트/음식점      1
84257        42148         음료      1
84258        42149       NULL      1

[84259 rows x 3 columns]

In [30]:
import pickle
with open("transaction_bhnl_h.pickle","wb") as fw:
    pickle.dump(transaction1010_h_df, fw)

In [31]:
transaction1010_h_df

Transaction       Item  Count
0                0   화장품/뷰티케어      1
1                1       패션잡화      1
2                1      퍼스널케어      1
3                1    테넌트/음식점      1
4                1  속옷/양말/홈웨어      1
...            ...        ...    ...
84254        42146       NULL      1
84255        42147       NULL      1
84256        42148    테넌트/음식점      1
84257        42148         음료      1
84258        42149       NULL      1

[84259 rows x 3 columns]

In [32]:
with open("C:/Users/tldms/jupyter/transaction_bhnl_h.pickle","rb") as fr:
    transaction1010_h_df=pickle.load(fr)

In [33]:
basket_h = transaction1010_h_df.pivot_table(index='Transaction', columns='Item', values='Count', aggfunc='sum').fillna(0)
basket_h.head()

Item         NULL   가구  건강식품  건강용품  건해산물  계절가전  공구/안전용품   과일   과자  구기/필드스포츠  \
Transaction                                                                   
0             0.0  0.0   0.0   0.0   0.0   0.0      0.0  0.0  0.0       0.0   
1             1.0  0.0   0.0   0.0   0.0   0.0      0.0  0.0  0.0       0.0   
2             0.0  0.0   0.0   0.0   0.0   0.0      0.0  0.0  0.0       0.0   
3             0.0  0.0   0.0   0.0   0.0   0.0      0.0  0.0  0.0       0.0   
4             0.0  0.0   0.0   0.0   0.0   0.0      0.0  0.0  0.0       0.0   

Item         ...  축산물  출산/육아용품  침구/수예  커피/차  컴퓨터  테넌트/음식점  패션잡화  퍼스널케어  \
Transaction  ...                                                         
0            ...  0.0      0.0    0.0   0.0  0.0      0.0   0.0    0.0   
1            ...  0.0      0.0    0.0   0.0  0.0      1.0   1.0    1.0   
2            ...  0.0      0.0    0.0   0.0  0.0      1.0   0.0    0.0   
3            ...  0.0      0.0    0.0   0.0  0.0      0.0   0.0    0.0   
4            ...  0.0      0.0    0.0   0.0  0.0      0.0   0.0    0.0   

Item         헬스/피트니스  화장품/뷰티케어  
Transaction                     
0                0.0       1.0  
1                0.0       0.0  
2                0.0       0.0  
3                0.0       1.0  
4                0.0       0.0  

[5 rows x 59 columns]

In [34]:
freq_item_h = apriori(basket_h, min_support = 0.01,use_colnames = True)
freq_item_h

C:\Users\tldms\anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:111: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support           itemsets
0   0.203915             (NULL)
1   0.012479             (건강식품)
2   0.014377             (건해산물)
3   0.066833               (과일)
4   0.090866               (과자)
5   0.012266         (구기/필드스포츠)
6   0.033405          (기타(비상품))
7   0.087117             (남성의류)
8   0.023867             (냉동식품)
9   0.057556             (냉장식품)
10  0.012242               (담배)
11  0.068588              (대용식)
12  0.020688          (문구/사무용품)
13  0.010107          (생활/주방가전)
14  0.013001            (세제/위생)
15  0.029822        (속옷/양말/홈웨어)
16  0.025196              (수산물)
17  0.108209            (스포츠패션)
18  0.025314          (식기/조리기구)
19  0.014567          (아웃도어/레저)
20  0.138482             (여성의류)
21  0.020214               (완구)
22  0.028873            (유아동의류)
23  0.051886              (유제품)
24  0.066002               (음료)
25  0.056323             (조리식품)
26  0.030368              (조미료)
27  0.034899               (주류)
28  0.020783             (주방잡화)
29  0.057509               (채소)
30  0.032574       (청소/세탁/욕실용품)
31  0.047117              (축산물)
32  0.015421            (침구/수예)
33  0.013855             (커피/차)
34  0.189110          (테넌트/음식점)
35  0.103962             (패션잡화)
36  0.023298            (퍼스널케어)
37  0.052835         (화장품/뷰티케어)
38  0.013903    (테넌트/음식점, NULL)
39  0.018149           (과일, 과자)
40  0.013737         (과일, 냉장식품)
41  0.015563          (과일, 대용식)
42  0.016584          (과일, 유제품)
43  0.023796           (과일, 채소)
44  0.010629   (과일, 청소/세탁/욕실용품)
45  0.014306          (과일, 축산물)
46  0.011103      (과일, 테넌트/음식점)
47  0.017224         (냉장식품, 과자)
48  0.020000          (대용식, 과자)
49  0.019288          (유제품, 과자)
50  0.018316           (음료, 과자)
51  0.010700          (조미료, 과자)
52  0.011364           (주류, 과자)
53  0.016512           (채소, 과자)
54  0.011340   (청소/세탁/욕실용품, 과자)
55  0.013571          (축산물, 과자)
56  0.017770      (테넌트/음식점, 과자)
57  0.017414      (남성의류, 스포츠패션)
58  0.019359       (남성의류, 여성의류)
59  0.019075    (남성의류, 테넌트/음식점)
60  0.013760       (남성의류, 패션잡화)
61  0.015777        (대용식, 냉장식품)
62  0.015065        (냉장식품, 유제품)
63  0.011483         (냉장식품, 음료)
64  0.017319         (냉장식품, 채소)
65  0.012456        (냉장식품, 축산물)
66  0.011198    (테넌트/음식점, 냉장식품)
67  0.015967         (대용식, 유제품)
68  0.013049          (대용식, 음료)
69  0.010913         (대용식, 조미료)
70  0.017011          (대용식, 채소)
71  0.010036  (대용식, 청소/세탁/욕실용품)
72  0.012123         (대용식, 축산물)
73  0.011222     (대용식, 테넌트/음식점)
74  0.010273          (채소, 수산물)
75  0.019193      (여성의류, 스포츠패션)
76  0.021756   (테넌트/음식점, 스포츠패션)
77  0.015018      (스포츠패션, 패션잡화)
78  0.024009    (테넌트/음식점, 여성의류)
79  0.019004       (여성의류, 패션잡화)
80  0.011886          (유제품, 음료)
81  0.016821          (채소, 유제품)
82  0.012313         (유제품, 축산물)
83  0.010320           (채소, 음료)
84  0.013926    (테넌트/음식점, 조리식품)
85  0.013974          (채소, 조미료)
86  0.012361   (청소/세탁/욕실용품, 채소)
87  0.020617          (채소, 축산물)
88  0.020996    (테넌트/음식점, 패션잡화)

In [35]:
freq_item_h.sort_values('support', ascending = False, inplace = True)
pd.set_option('display.max_rows', 500)
freq_item_h

support           itemsets
0   0.203915             (NULL)
34  0.189110          (테넌트/음식점)
20  0.138482             (여성의류)
17  0.108209            (스포츠패션)
35  0.103962             (패션잡화)
4   0.090866               (과자)
7   0.087117             (남성의류)
11  0.068588              (대용식)
3   0.066833               (과일)
24  0.066002               (음료)
9   0.057556             (냉장식품)
29  0.057509               (채소)
25  0.056323             (조리식품)
37  0.052835         (화장품/뷰티케어)
23  0.051886              (유제품)
31  0.047117              (축산물)
27  0.034899               (주류)
6   0.033405          (기타(비상품))
30  0.032574       (청소/세탁/욕실용품)
26  0.030368              (조미료)
15  0.029822        (속옷/양말/홈웨어)
22  0.028873            (유아동의류)
18  0.025314          (식기/조리기구)
16  0.025196              (수산물)
78  0.024009    (테넌트/음식점, 여성의류)
8   0.023867             (냉동식품)
43  0.023796           (과일, 채소)
36  0.023298            (퍼스널케어)
76  0.021756   (테넌트/음식점, 스포츠패션)
88  0.020996    (테넌트/음식점, 패션잡화)
28  0.020783             (주방잡화)
12  0.020688          (문구/사무용품)
87  0.020617          (채소, 축산물)
21  0.020214               (완구)
48  0.020000          (대용식, 과자)
58  0.019359       (남성의류, 여성의류)
49  0.019288          (유제품, 과자)
75  0.019193      (여성의류, 스포츠패션)
59  0.019075    (남성의류, 테넌트/음식점)
79  0.019004       (여성의류, 패션잡화)
50  0.018316           (음료, 과자)
39  0.018149           (과일, 과자)
56  0.017770      (테넌트/음식점, 과자)
57  0.017414      (남성의류, 스포츠패션)
64  0.017319         (냉장식품, 채소)
47  0.017224         (냉장식품, 과자)
70  0.017011          (대용식, 채소)
81  0.016821          (채소, 유제품)
42  0.016584          (과일, 유제품)
53  0.016512           (채소, 과자)
67  0.015967         (대용식, 유제품)
61  0.015777        (대용식, 냉장식품)
41  0.015563          (과일, 대용식)
32  0.015421            (침구/수예)
62  0.015065        (냉장식품, 유제품)
77  0.015018      (스포츠패션, 패션잡화)
19  0.014567          (아웃도어/레저)
2   0.014377             (건해산물)
45  0.014306          (과일, 축산물)
85  0.013974          (채소, 조미료)
84  0.013926    (테넌트/음식점, 조리식품)
38  0.013903    (테넌트/음식점, NULL)
33  0.013855             (커피/차)
60  0.013760       (남성의류, 패션잡화)
40  0.013737         (과일, 냉장식품)
55  0.013571          (축산물, 과자)
68  0.013049          (대용식, 음료)
14  0.013001            (세제/위생)
1   0.012479             (건강식품)
65  0.012456        (냉장식품, 축산물)
86  0.012361   (청소/세탁/욕실용품, 채소)
82  0.012313         (유제품, 축산물)
5   0.012266         (구기/필드스포츠)
10  0.012242               (담배)
72  0.012123         (대용식, 축산물)
80  0.011886          (유제품, 음료)
63  0.011483         (냉장식품, 음료)
52  0.011364           (주류, 과자)
54  0.011340   (청소/세탁/욕실용품, 과자)
73  0.011222     (대용식, 테넌트/음식점)
66  0.011198    (테넌트/음식점, 냉장식품)
46  0.011103      (과일, 테넌트/음식점)
69  0.010913         (대용식, 조미료)
51  0.010700          (조미료, 과자)
44  0.010629   (과일, 청소/세탁/욕실용품)
83  0.010320           (채소, 음료)
74  0.010273          (채소, 수산물)
13  0.010107          (생활/주방가전)
71  0.010036  (대용식, 청소/세탁/욕실용품)

In [38]:
rules_h = association_rules(freq_item_h, metric = "lift", min_threshold = 1)
rules_h.sort_values('support', ascending = False, inplace = True)
rules_h.head(20)

antecedents consequents  antecedent support  consequent support   support  \
0         (과일)        (채소)            0.066833            0.057509  0.023796   
1         (채소)        (과일)            0.057509            0.066833  0.023796   
2    (테넌트/음식점)     (스포츠패션)            0.189110            0.108209  0.021756   
3      (스포츠패션)   (테넌트/음식점)            0.108209            0.189110  0.021756   
4    (테넌트/음식점)      (패션잡화)            0.189110            0.103962  0.020996   
5       (패션잡화)   (테넌트/음식점)            0.103962            0.189110  0.020996   
6         (채소)       (축산물)            0.057509            0.047117  0.020617   
7        (축산물)        (채소)            0.047117            0.057509  0.020617   
8        (대용식)        (과자)            0.068588            0.090866  0.020000   
9         (과자)       (대용식)            0.090866            0.068588  0.020000   
11      (여성의류)      (남성의류)            0.138482            0.087117  0.019359   
10      (남성의류)      (여성의류)            0.087117            0.138482  0.019359   
12       (유제품)        (과자)            0.051886            0.090866  0.019288   
13        (과자)       (유제품)            0.090866            0.051886  0.019288   
14      (여성의류)     (스포츠패션)            0.138482            0.108209  0.019193   
15     (스포츠패션)      (여성의류)            0.108209            0.138482  0.019193   
16      (남성의류)   (테넌트/음식점)            0.087117            0.189110  0.019075   
17   (테넌트/음식점)      (남성의류)            0.189110            0.087117  0.019075   
18      (여성의류)      (패션잡화)            0.138482            0.103962  0.019004   
19      (패션잡화)      (여성의류)            0.103962            0.138482  0.019004   

    confidence      lift  leverage  conviction  
0     0.356053  6.191260  0.019952    1.463615  
1     0.413779  6.191260  0.019952    1.591835  
2     0.115042  1.063149  0.001292    1.007722  
3     0.201052  1.063149  0.001292    1.014947  
4     0.111027  1.067962  0.001336    1.007948  
5     0.201963  1.067962  0.001336    1.016105  
6     0.358498  7.608613  0.017907    1.485394  
7     0.437563  7.608613  0.017907    1.675727  
8     0.291595  3.209063  0.013768    1.283353  
9     0.220104  3.209063  0.013768    1.194277  
11    0.139798  1.604706  0.007295    1.061242  
10    0.222222  1.604706  0.007295    1.107666  
12    0.371742  4.091104  0.014574    1.447071  
13    0.212272  4.091104  0.014574    1.203605  
14    0.138599  1.280844  0.004208    1.035280  
15    0.177373  1.280844  0.004208    1.047278  
16    0.218954  1.157812  0.002600    1.038210  
17    0.100866  1.157812  0.002600    1.015290  
18    0.137228  1.319982  0.004607    1.038557  
19    0.182793  1.319982  0.004607    1.054223

## 결론 및 그에 따른 상한한도 결정하기

대분류를 기준으로 품목을 봤을 때 (테넌트/음식점을 제외하고) 의류라는 하나의 카테고리로 묶을 수 있었다<br>
(여성의류, 스포츠패션, 패션잡화, 남성의류)
그래서 의류라는 하나의 틀을 정의하고 그에 해당하는 모든 transaction을 기준으로 <br>
고객 당 한 번 구매 평균 찍어서(거래 횟수로 나누기) 금액을 구하고 그 금액보다 높은 금액으로 기준으로 상한한도를 결정하겠다.

In [42]:
import cx_Oracle
import os

LOCATION = r"C:\instantclient_21_6"
os.environ["PATH"] = LOCATION + ";" + os.environ["PATH"]

In [43]:
connect = cx_Oracle.connect("system", "sieun257", "localhost:1521/orcl")
cursor = connect.cursor()

중분류에서 support를 기준으로 의류와 관련된 상위 품목인 것들의 대분류 카테고리를 알아보겠다

In [44]:
clac = pd.read_sql("select * from LPAY_PD_CLAC" , con = connect)
list = ['여성의류상의','여성의류아우터','남성의류상의','남성의류세트','남성일반스포츠의류','남성스포츠화', '여성의류세트']
clac = clac.drop(['PD_C','PD_NM'],axis = 1)
for i in list:
    a = clac['CLAC_MCLS_NM']==i
    print(clac[a].drop_duplicates(),'\n')
# 대분류 파악

    CLAC_HLV_NM CLAC_MCLS_NM
948        여성의류       여성의류상의 

    CLAC_HLV_NM CLAC_MCLS_NM
957        여성의류      여성의류아우터 

    CLAC_HLV_NM CLAC_MCLS_NM
289        남성의류       남성의류상의 

    CLAC_HLV_NM CLAC_MCLS_NM
296        남성의류       남성의류세트 

    CLAC_HLV_NM CLAC_MCLS_NM
779       스포츠패션    남성일반스포츠의류 

    CLAC_HLV_NM CLAC_MCLS_NM
776       스포츠패션       남성스포츠화 

    CLAC_HLV_NM CLAC_MCLS_NM
955        여성의류       여성의류세트 



C:\Users\tldms\anaconda3\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


유의미한 중분류를 기준으로 봤을 때 대분류는 여성의류, 남성의류, 스포츠패션이다.<br>
그것을 포함하고 대분류에서의 연관관계를 포함해서 의류라는 하나의 틀의 기준을 세우겠다.

이제부터 앞서 정한 기준으로 큰손nl에 대한 전략을 세울 것이다.<br>
큰손nl의 transaction에서 한 번 결제할 때의 평균 구매금액을 계산해보자

In [45]:
transaction1010

CUST  MA_FEM_DV  AGES  C_ZON_HLV  BUY_AM  COP_C  CHNL_DV  PD_C  \
0       M117855953          0    30         11  759500    110        1   949   
1       M117855953          0    30         11  167000    110        1   816   
2       M117855953          0    30         11   48000    110        1   816   
3       M117855953          0    30         11  126500    110        1   959   
4       M117855953          0    30         11   19700    110        1  1227   
...            ...        ...   ...        ...     ...    ...      ...   ...   
131135  M058847456          0    40         10    5400    410        1  4000   
131136  M058847456          0    40         10   15700    410        1  4000   
131137  M058847456          0    40         10   15300    410        1  4000   
131138  M058847456          0    40         10    9300    410        1  4000   
131139  M058847456          0    40         10   20800    410        1  4000   

        BUY_CT  S_ZON_HLV       DE_DT  MONTH  DATE  DAY  DE_HR CLAC_HLV_NM  \
0            1         11  2021-01-01      1     1    4     11        여성의류   
1            1         11  2021-01-13      1    13    2     20       스포츠패션   
2            1         11  2021-01-24      1    24    6     13       스포츠패션   
3            1         11  2021-02-14      2    14    6     18        여성의류   
4            1         11  2021-02-14      2    14    6     20       자동차용품   
...        ...        ...         ...    ...   ...  ...    ...         ...   
131135       1         11  2021-08-14      8    14    5     12        NULL   
131136       1         11  2021-10-14     10    14    3     16        NULL   
131137       1         11  2021-11-09     11     9    1     16        NULL   
131138       1         11  2021-09-24      9    24    4     10        NULL   
131139       1         11  2021-04-20      4    20    1     16        NULL   

       CLAC_MCLS_NM       PD_NM  LPAY_USE  
0            여성의류상의     여성티셔츠/탑         0  
1            여성스포츠화  여성런닝/트레이닝화         0  
2            여성스포츠화  여성런닝/트레이닝화         0  
3           여성의류아우터        여성재킷         0  
4          액세서리편의용품     수납/정리용품         0  
...             ...         ...       ...  
131135         NULL        NULL         0  
131136         NULL        NULL         0  
131137         NULL        NULL         0  
131138         NULL        NULL         0  
131139         NULL        NULL         0  

[131140 rows x 19 columns]

In [46]:
condition = transaction1010['CLAC_HLV_NM'].isin(['여성의류', '남성의류', '스포츠패션'])
transaction1010_isin = transaction1010.loc[condition]
transaction1010_isin

CUST  MA_FEM_DV  AGES  C_ZON_HLV  BUY_AM  COP_C  CHNL_DV  PD_C  \
0       M117855953          0    30         11  759500    110        1   949   
1       M117855953          0    30         11  167000    110        1   816   
2       M117855953          0    30         11   48000    110        1   816   
3       M117855953          0    30         11  126500    110        1   959   
5       M117855953          0    30         11  372370    110        1   949   
...            ...        ...   ...        ...     ...    ...      ...   ...   
131125  M058847456          0    40         10  122600    160        2   777   
131126  M058847456          0    40         10   36800    160        2   817   
131128  M058847456          0    40         10   59000    160        2   969   
131129  M058847456          0    40         10   39200    160        2   290   
131132  M058847456          0    40         10   26000    160        2   954   

        BUY_CT  S_ZON_HLV       DE_DT  MONTH  DATE  DAY  DE_HR CLAC_HLV_NM  \
0            1         11  2021-01-01      1     1    4     11        여성의류   
1            1         11  2021-01-13      1    13    2     20       스포츠패션   
2            1         11  2021-01-24      1    24    6     13       스포츠패션   
3            1         11  2021-02-14      2    14    6     18        여성의류   
5            1         11  2021-02-14      2    14    6     20        여성의류   
...        ...        ...         ...    ...   ...  ...    ...         ...   
131125       1          0  2021-05-22      5    22    5      6       스포츠패션   
131126       1          0  2021-05-22      5    22    5      7       스포츠패션   
131128       1          0  2021-05-29      5    29    5     11        여성의류   
131129       1          0  2021-07-02      7     2    4      4        남성의류   
131132       1          0  2021-12-01     12     1    2     16        여성의류   

       CLAC_MCLS_NM        PD_NM  LPAY_USE  
0            여성의류상의      여성티셔츠/탑         0  
1            여성스포츠화   여성런닝/트레이닝화         0  
2            여성스포츠화   여성런닝/트레이닝화         0  
3           여성의류아우터         여성재킷         0  
5            여성의류상의      여성티셔츠/탑         0  
...             ...          ...       ...  
131125       남성스포츠화   남성런닝/트레이닝화         0  
131126       여성스포츠화  여성스포츠샌들/슬리퍼         0  
131128       여성의류하의        여성청바지         0  
131129       남성의류상의        남성티셔츠         0  
131132       여성의류상의    여성스웨터/풀오버         0  

[26344 rows x 19 columns]

조건에 맞는 26344개의 transaction을 찾았다

한번 거래기록의 평균을 보려면 일단 한번 거래할때의 구매금액을 알아야한다<br>
이 한번 거래할때의 기준을 '한 사람'이 '몇 월', '몇 일', '몇 시'로 기준을 세우면 한번 거래 기록이 될 것이다.<br>
즉 (column) cust, month, date, day, de_hr을 기준으로 groupby를 하고 그때의 value값을 buy_am으로 sum하면 한번 거래할때의 구매금액이 될것이다.<br>

In [47]:
one_transaction = transaction1010_isin.groupby(['CUST', 'MONTH','DATE','DAY','DE_HR']).agg({'BUY_AM': np.sum})
one_transaction = one_transaction.reset_index()
one_transaction

CUST  MONTH  DATE  DAY  DE_HR   BUY_AM
0      M000261625      4     3    5     13   179000
1      M000261625     10    11    0     14   588000
2      M000350564      2    25    3     19   278600
3      M000350564      5     1    5     13  2224000
4      M000350564      5    28    4     12   317300
...           ...    ...   ...  ...    ...      ...
14715  M997843584     10    10    6     14  1093000
14716  M997843584     10    10    6     15    79000
14717  M997843584     10    13    2     20    89000
14718  M997843584     10    17    6     12   138000
14719  M997843584     11     7    6     13   359760

[14720 rows x 6 columns]

In [48]:
one_transaction['BUY_AM'].mean() 

285440.48369565216

한번 평균 금액은 28,5440원이다.